### PR TITLE
Multi validator support 

### LINKED ISSUE #55: Support multi validators share
Once the node will work with storage we need to support multi validators share storage based. #19

### PR DESCRIPTION
**Pull request by nivBlox**
resolve #55 
Fetch all validators from db and for each create validator wrapper that managing the ibfts instances.

- Load validator from .env
- listen to events from smart contract for added new validator (will be implemented in  #52)

### CODE REVIEW

File: network/p2p/p2p.go [hunk 1]
```diff
@@ -47,25 +49,27 @@ type listener struct {
...
+	logger           *zap.Logger
+	privKey          *ecdsa.PrivateKey
+	peers            *peers.Status
+	host             p2pHost.Host
+	pubsub           *pubsub.PubSub
 }
 
 // New is the constructor of p2pNetworker
-func New(ctx context.Context, logger *zap.Logger, cfg *Config) (network.Network, error) {
+func New(ctx context.Context, logger *zap.Logger, validatorStorage collections.IValidator, cfg *Config) (network.Network, error) {
```
Comment (alonmuroch) [inline network/p2p/p2p.go]: no need to have the storage in the struct if you only use it to generate topics. 
BTW, the way it's setup it's not dynamic while reading from the contract. You'd have to restart the node to apply new validators?

File: cli/node.go [hunk 2]
```diff
@@ -143,32 +141,21 @@ var startNodeCmd = &cobra.Command{
...
+		network, err := p2p.New(cmd.Context(), logger, &validatorStorage, &cfg)
 		if err != nil {
 			logger.Fatal("failed to create network", zap.Error(err))
 		}
 
-		msgQ := msgqueue.New()
-
 		ssvNode := node.New(node.Options{
 			ValidatorStorage: validatorStorage,
+			IbftStorage:      ibftStorage,
```
Comment (alonmuroch) [inline cli/node.go]: You only use it to setup validators, no need to pass pointer to db or have it in the struct.. it makes the code more spaghetti like.

File: docker-compose.yaml [hunk 3]
```diff
@@ -90,8 +89,8 @@ services:
   ssv-node-4-dev:
     << : *default-dev
     container_name: ssv-node-4-dev
-    ports:
-      - 40008:40008
+#    ports:
```
Comment (alonmuroch) [inline docker-compose.yaml:92]: is the commenting out by purpose?

File: storage/collections/validator_storage.go [hunk 4]
```diff
@@ -19,10 +19,10 @@ type IValidator interface {
 
 // Validator model for ValidatorStorage struct creation
 type Validator struct {
-	NodeID     uint64
-	PubKey     *bls.PublicKey
-	ShareKey   *bls.SecretKey
-	Committiee map[uint64]*proto.Node
+	NodeID    uint64
+	PubKey    *bls.PublicKey
```
Comment (alonmuroch) [inline storage/collections/validator_storage.go]: this should be validatorPK right?

File: ibft/instance.go [hunk 5]
```diff
@@ -28,6 +28,7 @@ type InstanceOptions struct {
 	ValueCheck     valcheck.ValueCheck
 	LeaderSelector leader.Selector
 	Params         *proto.InstanceParams
+	PublicKey      []byte
```
Comment (alonmuroch) [inline ibft/instance.go]: no need to add it, we already have validatorPK

File: network/p2p/p2p.go [hunk 6]
```diff
@@ -47,25 +49,27 @@ type listener struct {
...
-	listenersLock sync.Locker
-	dv5Listener   iListener
-	listeners     []listener
-	logger        *zap.Logger
-	privKey       *ecdsa.PrivateKey
-	peers         *peers.Status
-	host          p2pHost.Host
-	pubsub        *pubsub.PubSub
+	ctx              context.Context
+	validatorStorage collections.IValidator
```
Comment (alonmuroch) [inline network/p2p/p2p.go]: no need to have the storage in the struct if you only use it to generate topics.
BTW, the way it's setup it's not dynamic while reading from the contract. You'd have to restart the node to apply new validators?

File: network/p2p/p2p.go [hunk 7]
```diff
@@ -103,7 +107,7 @@ func New(ctx context.Context, logger *zap.Logger, cfg *Config) (network.Network,
 		return nil, errors.New("Unsupported discovery flag")
 	}
 
-	n.logger = logger.With(zap.String("id", n.host.ID().String()), zap.String("Topic", n.cfg.TopicName))
+	n.logger = logger.With(zap.String("id", n.host.ID().String()))
```
Comment (alonmuroch) [inline network/p2p/p2p.go:110]: we need our logs to be multi validator as well

File: network/p2p/p2p_decided.go [hunk 8]
```diff
@@ -18,7 +19,13 @@ func (n *p2pNetwork) BroadcastDecided(msg *proto.SignedMessage) error {
 		return errors.Wrap(err, "failed to marshal message")
 	}
 
-	return n.cfg.Topic.Publish(n.ctx, msgBytes)
+	topic, err := n.GetTopic(msg)
+	if err != nil {
+		return errors.Wrap(err, "failed to get topic")
+	}
+
+	n.logger.Debug("Broadcasting to topic", zap.Any("topic", topic), zap.Any("peers", topic.ListPeers()))
```
Comment (alonmuroch) [inline network/p2p/p2p_decided.go:27]: no sure we need this log

File: network/p2p/p2p_ibft.go [hunk 9]
```diff
@@ -19,8 +19,13 @@ func (n *p2pNetwork) Broadcast(msg *proto.SignedMessage) error {
...
 	}
 
-	n.logger.Debug("Broadcasting to topic", zap.Any("topic", n.cfg.Sub.Topic()), zap.Any("peers", n.cfg.Topic.ListPeers()))
-	return n.cfg.Topic.Publish(n.ctx, msgBytes)
+	topic, err := n.GetTopic(msg)
+	if err != nil {
+		return errors.Wrap(err, "failed to get topic")
+	}
+
+	n.logger.Debug("Broadcasting to topic", zap.Any("topic", topic), zap.Any("peers", topic.ListPeers()))
```
Comment (alonmuroch) [inline network/p2p/p2p_ibft.go:27]: not sure we need this log

File: network/p2p/p2p_signatures.go [hunk 10]
```diff
@@ -17,8 +18,13 @@ func (n *p2pNetwork) BroadcastSignature(msg *proto.SignedMessage) error {
 	if err != nil {
 		return errors.Wrap(err, "failed to marshal message")
 	}
+	topic, err := n.GetTopic(msg)
+	if err != nil {
+		return errors.Wrap(err, "failed to get topic")
+	}
 
-	return n.cfg.Topic.Publish(n.ctx, msgBytes)
+	n.logger.Debug("Broadcasting to topic", zap.Any("topic", topic), zap.Any("peers", topic.ListPeers()))
```
Comment (alonmuroch) [inline network/p2p/p2p_signatures.go:26]: not sure we need this log

File: network/p2p/utils_test.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
+package p2p
```
Comment (alonmuroch) [inline network/p2p/utils_test.go]: I think we called it test_utils up until now so we should be consistent

File: node/node.go [hunk 12]
```diff
@@ -59,13 +57,12 @@ type ssvNode struct {
 func New(opts Options) Node {
 	return &ssvNode{
 		validatorStorage:           opts.ValidatorStorage,
+		ibftStorage:                opts.IbftStorage,
```
Comment (alonmuroch) [inline node/node.go:66]: see my comment about why we don't need to have ibft storage instance

File: node/node.go [hunk 13]
```diff
@@ -110,20 +109,18 @@ func (n *ssvNode) Start(ctx context.Context) error {
...
-						ValidatorPK: validator.PubKey,
-						ShareSK:     validator.ShareKey,
-						Committee:   validator.Committiee,
-					}
-
-					// execute task if slot already began
-					if slot == uint64(n.getCurrentSlot()) {
+					// execute task if slot already began and not pass 1 epoch
+					currentSlot := uint64(n.getCurrentSlot())
+					if slot >= currentSlot && slot - currentSlot <= 12 {
```
Comment (alonmuroch) [inline node/node.go]: why 12? and if so why not in a param?

File: validator/duty_execution.go [hunk 14]
```diff
@@ -81,57 +80,51 @@ func (n *ssvNode) waitForSignatureCollection(logger *zap.Logger, identifier []by
...
 	if err != nil {
 		return errors.Wrap(err, "failed to sign input data")
 	}
 
 	// TODO - should we construct it better?
-	if err := n.network.BroadcastSignature(&proto.SignedMessage{
+	if err := v.network.BroadcastSignature(&proto.SignedMessage{
 		Message: &proto.Message{
-			Lambda: identifier,
+			Lambda:      identifier,
```
Comment (alonmuroch) [inline validator/duty_execution.go:93]: seq_number is missing

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: network/p2p/p2p.go (hunk 1 above)
Comment (nivBlox) [reply network/p2p/p2p.go]: Will  remove it from p2p and add "addTopic" interface func. Validator wrapper will call this func when created and in that way we will manage every new validator topic

File: cli/node.go (hunk 2 above)
Comment (nivBlox) [reply cli/node.go]: needed for init setup validators (already saved validator in db) we can pass array of validators but i think its much messy that current solution. @lior-blox pls add your opinion about it too

File: docker-compose.yaml (hunk 3 above)
Comment (nivBlox) [reply docker-compose.yaml:92]: yes, need to figure it out. @lior-blox what do you say?

File: storage/collections/validator_storage.go (hunk 4 above)
Comment (nivBlox) [reply storage/collections/validator_storage.go]: right, will change it

File: ibft/instance.go (hunk 5 above)
Comment (nivBlox) [reply ibft/instance.go]: with that we populating the state pk

File: network/p2p/p2p_signatures.go (hunk 10 above)
Comment (nivBlox) [reply network/p2p/p2p_signatures.go:26]: agree, will remove it

File: network/p2p/utils_test.go (hunk 11 above)
Comment (nivBlox) [reply network/p2p/utils_test.go]: agree, will change it

File: node/node.go (hunk 13 above)
Comment (nivBlox) [reply node/node.go]: to support up to 2 epochs. do you think need differ value? 
Will add flag

File: validator/duty_execution.go (hunk 14 above)
Comment (nivBlox) [reply validator/duty_execution.go:93]: missing on stage too. what value should be pass?

File: network/p2p/p2p_signatures.go (hunk 10 above)
Comment (nivBlox) [reply network/p2p/p2p_signatures.go:26]: @alonmuroch on sec thought, it is important for debugging

File: ibft/instance.go (hunk 5 above)
Comment (alonmuroch) [reply ibft/instance.go]: but we already have validator PK it's just a duplicate

File: validator/validator.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+	SignatureCollectionTimeout time.Duration
+	SlotQueue                  slotqueue.Queue
+}
+
+// Validator struct that manages all ibft wrappers
+type Validator struct {
+	ctx                        context.Context
+	logger                     *zap.Logger
+	ValidatorShare             *collections.Validator
+	ibftStorage                collections.IbftStorage
```
Comment (alonmuroch) [inline validator/validator.go]: this should be an interface not the struct

File: validator/duty_execution.go (hunk 14 above)
Comment (alonmuroch) [reply validator/duty_execution.go:93]: please open an issue for that for later

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: node/node.go (hunk 13 above)
Comment (alonmuroch) [reply node/node.go]: please add comment and fix to 1 epoch

File: validator/duty_execution.go (hunk 14 above)
Comment (nivBlox) [reply validator/duty_execution.go:93]: #65

File: cli/flags/node.go [hunk 16]
```diff
@@ -106,6 +107,16 @@ func GetSignatureCollectionTimeValue(c *cobra.Command) (time.Duration, error) {
 	return time.Second * time.Duration(v), nil
 }
 
+// AddDutySlotsLimit adds the duty slots limit flag to the command
+func AddDutySlotsLimit(c *cobra.Command) {
+	cliflag.AddPersistentIntFlag(c, dutySlotsLimit, 32, "Duties max slots delay to attest", false)
```
Comment (alonmuroch) [inline cli/flags/node.go:112]: explanation is really unclear here..

File: network/network.go [hunk 17]
```diff
@@ -58,4 +59,6 @@ type Network interface {
 
 	// ReceivedSyncMsgChan returns the channel for sync messages
 	ReceivedSyncMsgChan() <-chan *SyncChanObj
+
+	SubscribeTopic(validatorPk *bls.PublicKey) error
```
Comment (alonmuroch) [inline network/network.go]: not sure this should be part of an interface.. topics are really implementation specific

File: network/p2p/config.go [hunk 18]
```diff
@@ -14,7 +14,6 @@ type Config struct {
 	HostAddress         string
 	HostDNS             string
 	HostID              peer.ID
-	TopicName           string
-	Topic               *pubsub.Topic
-	Sub                 *pubsub.Subscription
+	Topics map[string]*pubsub.Topic
```
Comment (alonmuroch) [inline network/p2p/config.go:17]: any reason we need to hold topics and pubsubs in the struct? isn't just holding pubsub is enough?

File: network/p2p/p2p.go (hunk 7 above)
Comment (alonmuroch) [reply network/p2p/p2p.go:110]: @nivBlox

File: network/p2p/p2p.go [hunk 19]
```diff
@@ -172,74 +176,96 @@ func New(ctx context.Context, logger *zap.Logger, cfg *Config) (network.Network,
...
 	}
+	topic := pk.SerializeToHexStr()
+	if _, ok := n.cfg.Topics[topic]; !ok {
+		return nil, errors.New("topic is not exist or registered")
+	}
+	return n.cfg.Topics[topic], nil
 }
 
+
 func getTopic(topicName string) string {
```
Comment (alonmuroch) [inline network/p2p/p2p.go]: @nivBlox we need this function?

File: node/node.go (hunk 12 above)
Comment (alonmuroch) [reply node/node.go:66]: @nivBlox

File: storage/collections/validator_storage.go [hunk 20]
```diff
@@ -35,7 +35,7 @@ type ValidatorStorage struct {
 // NewValidator creates new validator storage
 func NewValidator(db storage.Db, logger *zap.Logger) ValidatorStorage {
```
Comment (alonmuroch) [inline storage/collections/validator_storage.go]: this should be called NewValidatorStorage

File: storage/collections/validator_storage.go [hunk 21]
```diff
@@ -101,7 +110,7 @@ func (v *Validator) Serialize() ([]byte, error) {
 	value := serializedValidator{
```
Comment (alonmuroch) [inline storage/collections/validator_storage.go]: no tests

File: storage/collections/validator_storage.go [hunk 22]
```diff
@@ -128,9 +137,9 @@ func (v *Validator) Deserialize(obj storage.Obj) (*Validator, error) {
 		return nil, errors.Wrap(err, "Failed to get pubkey")
```
Comment (alonmuroch) [inline storage/collections/validator_storage.go:137]: no tests

File: storage/kv/badger.go [hunk 23]
```diff
@@ -63,19 +63,23 @@ func (b *BadgerDb) GetAllByBucket(prefix []byte) ([]storage.Obj, error) {
...
 		opt := badger.DefaultIteratorOptions
+		opt.Prefix = prefix
 		it := txn.NewIterator(opt)
 		defer it.Close()
 		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
 			item := it.Item()
 			resKey := item.Key()
 			trimmedResKey := bytes.TrimPrefix(resKey, prefix)
 			val, err := item.ValueCopy(nil)
+			if err != nil {
```
Comment (alonmuroch) [inline storage/kv/badger.go:87]: @nivBlox we got no tests for this whole db

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: network/network.go (hunk 17 above)
Comment (nivBlox) [reply network/network.go]: what you mean is not clear for me, how you suggest to implement that func (or where?)

File: network/p2p/config.go (hunk 18 above)
Comment (nivBlox) [reply network/p2p/config.go:17]: we need the topics for the broadcasting, pubsub is the gossip

File: network/p2p/p2p.go (hunk 19 above)
Comment (nivBlox) [reply network/p2p/p2p.go]: Yes, each validator need to fetch the proper topic in order to broadcast msg

File: network/network.go (hunk 17 above)
Comment (alonmuroch) [reply network/network.go]: Maybe rename it to SubscribeToValidatorNetwork + add comment for the public function

File: network/p2p/config.go [hunk 24]
```diff
@@ -14,7 +14,6 @@ type Config struct {
...
+	Subs   []*pubsub.Subscription
```
Comment (alonmuroch) [inline network/p2p/config.go:18]: please open an issue to decide if we want to remove it

File: network/p2p/p2p.go [hunk 25]
```diff
@@ -172,74 +176,96 @@ func New(ctx context.Context, logger *zap.Logger, cfg *Config) (network.Network,
...
+							n.logger.Error("received unsupported message", zap.Int32("msg type", int32(cm.Type)))
+						}
+					}(ls)
+				}
 			}
 		}
+	}(sub)
+}
+
+func (n *p2pNetwork) GetTopic(msg *proto.SignedMessage) (*pubsub.Topic, error) {
```
Comment (alonmuroch) [inline network/p2p/p2p.go:254]: @nivBlox please add comments for public functions

File: network/p2p/p2p.go (hunk 19 above)
Comment (alonmuroch) [reply network/p2p/p2p.go]: change name

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: network/p2p/config.go (hunk 24 above)
Comment (nivBlox) [reply network/p2p/config.go:18]: https://github.com/ethereum/eth2-ssv/issues/66

**Pull request merged**